mysql-test/include/search_pattern_in_file.inc

```diff
@@ -71,54 +59,64 @@ perl;
     my @search_files= glob($ENV{SEARCH_FILE});
     my $search_pattern= $ENV{SEARCH_PATTERN} or die "SEARCH_PATTERN not set";
     my $search_range=   $ENV{SEARCH_RANGE};
-    my $content;
-    foreach my $search_file (@search_files) {
-        open(FILE, '<', $search_file) || die("Can't open file $search_file: $!");
-        my $file_content;
-        if ($search_range > 0) {
-           read(FILE, $file_content, $search_range, 0);
-        } elsif ($search_range < 0) {
-           my $size= -s $search_file;
-           $search_range = -$size if $size > -$search_range;
-           seek(FILE, $search_range, 2);
-           read(FILE, $file_content, -$search_range, 0);
-        } else {
-          while(<FILE>) { # error log
-            if (/^CURRENT_TEST:/) {
-              $content='';
+    my $timeout=        $ENV{SEARCH_TIMEOUT} || 60;
+    my @matches;
+    my $res;
+
+    my $start_time= time();
+    for (;;) {
+        my $content;
+        foreach my $search_file (@search_files) {
+            open(FILE, '<', $search_file) || die("Can't open file $search_file: $!");
+            my $file_content;
+            if ($search_range > 0) {
+               read(FILE, $file_content, $search_range, 0);
+            } elsif ($search_range < 0) {
+               my $size= -s $search_file;
+               $search_range = -$size if $size > -$search_range;
+               seek(FILE, $search_range, 2);
+               read(FILE, $file_content, -$search_range, 0);
             } else {
-              $content.=$_;
+              while(<FILE>) { # error log
+                if (/^CURRENT_TEST:/) {
+                  $content='';
+                } else {
+                  $content.=$_;
+                }
+              }
+            }
+            close(FILE);
+            $content.= $file_content;
+        }
+        @matches= ($content =~ /$search_pattern/gs);
+        $res=@matches ? "FOUND " . scalar(@matches) : "NOT FOUND";
+
+        if ($ENV{SEARCH_WAIT} and not $res =~ /^$ENV{SEARCH_WAIT}/) {
```

Comment:
I think there is a bug here. If $res is "FOUND 10" and SEARCH_WAIT is "FOUND 1", this will terminate the wait (but "FOUND 9" will not for example).

More generally, it is not a good idea to interpolate user input directly into a regexp (unless specifically intended as a user regexp search facility), it is prone to problems like this. Probably better to provide a separate SEARCH_COUNT (for example) parameter, which specifies a specific count to wait for or abort on, if given, that is simpler and more robust.

Reply:
On the contrary, when I suggested something similar earlier (for `wait_for_pattern_in_file.inc`) I thought that this provides the desirable flexibility. One can search for `/FOUND 1\b/` or for `/FOUND 10\b/` or for `/FOUND [4-9]/` etc.

Normally `/FOUND 1/` will work though. I can invent a rather artificial case where one needs to wait for "1" but keep waiting if he number is "10", but i don't think it'll happen practically. And if it will, one can always add `\b`

Reply:
Sure, this is fine too, this is "specifically intended as a user regexp search facility".

So then just update the file comments to explain that SEARCH_WAIT is a regexp that waits until the search result (which will be "FOUND <n>" or "NOT FOUND") matches that regexp.

And also document that the regexp will be prepended with start-of-string-anchor "^".